### ISSUE TITLE
Spurious Dequantize/Cast/Quantize sequence of ops at the end of a QAT TFLite model.

### ISSUE THREAD

**Issue by corvoysier** [tensorflow/model-optimization#431]
**Describe the bug**
When converting a MNIST model trained using QAT, there is a spurious Dequantize/Cast/Quantize sequence of ops at the end of the TFLite model.

**System information**

TensorFlow version (installed from source or binary): TF nightly 2.3.0-dev20200616

TensorFlow Model Optimization version (installed from source or binary): 0.3.0

Python version: 3.6.9

**Describe the expected behavior**

Using the same workflow as the QAT MNIST example, produce a TFLite model from a slightly more complex model. 

As with the tutorial model (see tutorial_integer.png), after the conversion to TFLite, the converted model should have the following structure:

- input (float),
- quantize,
- several standard ops with integer tensors,
- dequantize,
- identity.

**Describe the current behavior**

The generated model contains a spurious dequantize/cast/quantize sequence before the final dequantize (see mnist_integer.png). This doesn't seem to affect the accuracy, but makes the model incompatible with the Edge TPU compiler.

**Code to reproduce the issue**

The training script to produce the model is in that gist:

https://gist.github.com/corvoysier/e4d9c6b96b103928544a57054917aace

**Screenshots**

![tutorial_integer](https://user-images.githubusercontent.com/58857004/85004489-e88c4780-b157-11ea-8021-e8e8f8baedbe.png)
![mnist_integer](https://user-images.githubusercontent.com/58857004/85004498-eb873800-b157-11ea-98c0-dc5d8cab7599.png)

**Comment by suharshs** [comment]
Hi, we are actively working on integer input and output support in the conversion for QAT. (The lack of this support is why you are getting spurious dequantize and quantize operations that edgetpu doesn't like). We will update this issue when that is ready.

If this is blocking you, we have a workaround script that can change the input and output types for you to uint8, which EdgeTPU requires: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/python/modify_model_interface.py

Hope that helps!

**Comment by Namburger** [comment]
@corvoysier @suharshs 
FYI, with `2.3.0-dev20200618`, my model with the exact same code as above looks very different. 
![Screenshot 2020-06-18 at 11 28 55 AM](https://user-images.githubusercontent.com/25712149/85047037-ea094000-b156-11ea-91bc-cd307c56ac6a.png)

So I set the tensor shapes to be static before converting is giving me better results:

```
q_aware_model.input.set_shape((1,) + q_aware_model.input.shape[1:])
    return converter.convert()
```
![Screenshot 2020-06-18 at 11 45 03 AM](https://user-images.githubusercontent.com/25712149/85048750-63099700-b159-11ea-92d5-0265277b164d.png)

Accuracy is still good:
```
Baseline test accuracy: 0.9463000297546387
Quantized float test accuracy: 0.9865999817848206
Integer test accuracy: 0.9865
```
and it compiles:
```
edgetpu_compiler -s mnist_integer.tflite                                  vunam@penguin
Edge TPU Compiler version 2.1.302470888

Model compiled successfully in 76 ms.

Input model: mnist_integer.tflite
Input size: 332.88KiB
Output model: mnist_integer_edgetpu.tflite
Output size: 537.03KiB
On-chip memory used for caching model parameters: 130.50KiB
On-chip memory remaining for caching model parameters: 7.77MiB
Off-chip memory used for streaming uncached model parameters: 392.06KiB
Number of Edge TPU subgraphs: 1
Total number of operations: 10
Operation log: mnist_integer_edgetpu.log

Model successfully compiled but not all operations are supported by the Edge TPU. A percentage of the model will instead run on the CPU, which is slower. If possible, consider updating your model to use only operations supported by the Edge TPU. For details, visit g.co/coral/model-reqs.
Number of operations that will run on Edge TPU: 8
Number of operations that will run on CPU: 2

Operator                       Count      Status

CONV_2D                        4          Mapped to Edge TPU
RESHAPE                        1          Mapped to Edge TPU
DEQUANTIZE                     1          Operation is working on an unsupported data type
FULLY_CONNECTED                1          Mapped to Edge TPU
MAX_POOL_2D                    2          Mapped to Edge TPU
QUANTIZE                       1          Operation is otherwise supported, but not mapped due to some unspecified limitation
```

cheers!

**Comment by corvoysier** [comment]
@Namburger

> FYI, with `2.3.0-dev20200618`, my model with the exact same code as above looks very different.

Are you sure the first model you obtain is from the nightly ? What you got is exactly what I had when I used TF 2.2 release, and it was actually a float quantized model (you can check the tensors).

> So I set the tensor shapes to be static before converting is giving me better results:
> `q_aware_model.input.set_shape((1,) + q_aware_model.input.shape[1:])`

Yes, that did the trick for me also. I now have the correct sequence.

**Comment by Namburger** [comment]
Humm I may have made a mistake of using tf2.2 but I remember specifically checking tf version before trying your code. Not 💯 on this

**Comment by LLNLanLeN** [comment]
@Namburger I second what @corvoysier  said. I was trying out 2 difference version on my side. TF nightly give a clean graphs compare to TF 2.2. I believe I mentioned somewhere in documents that it's recommended to use optimization toolkit with TF nightly

**Comment by Namburger** [comment]
@LLNLanLeN may I did made that mistake lol

**Comment by LLNLanLeN** [comment]
@Namburger  may be, haha. But the graphs turns out very well. I have a quick question actually, the script. Were you able to run this script as recommended above? 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/python/modify_model_interface.py

I install the optimization tookit via pip, but when I run the command, it couldn't locate the file (I couldn't locate it in my environment either)

**Comment by Namburger** [comment]
No, by modifying the script to set [static input shape](https://github.com/tensorflow/model-optimization/issues/431#issuecomment-646163906), I didn't have any troubles with the compiler so I didn't tried that work around
But from the [build file](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/python/BUILD#L11) it looks like it should be something like this to run (from tensorflow root):
```
bazel run -c opt //tensorflow/lite/tools/optimize/python:modify_python_inferace -- --input model ...
```

**Comment by LLNLanLeN** [comment]
@Namburger I see, thank you. I'll try the script out with bazel

**Comment by MeghnaNatraj** [comment]
@LLNLanLeN As @Namburger pointed out, you should be able to modify the model input and output by running
```
$ bazel run -c opt //tensorflow/lite/tools/optimize/python:modify_python_inferace -- --input_file=<path/to/quant_model.tflite> --output_file=<path/to/quant_model_int_io_type.tflite> --input_type=INT8 --output_type=INT8
```

You can change 'INT8' to 'UINT8' as well.

If you are using `tf-nightly`, you can directly modify the input and output type of quantized models during conversion, by using this code: https://www.tensorflow.org/lite/performance/post_training_quantization#integer_only

The exception to this is quantize-aware trained models. If the model is quantize-aware trained, then in order to modify the input and output type, you need to first convert the model to a quantized tflite model (which will always have float input and ouytput) and then use the `bazel` command provided above to modify the input and output type to `tf.int8` ot `tf.int8` .

**Comment by LLNLanLeN** [comment]
@MeghnaNatraj  thanks for confirming. I've managed to use the script

**Comment by aa12356jm** [comment]
mark

**Comment by EscVM** [comment]
> @LLNLanLeN As @Namburger pointed out, you should be able to modify the model input and output by running
> 
> ```
> $ bazel run -c opt //tensorflow/lite/tools/optimize/python:modify_python_inferace -- --input_file=<path/to/quant_model.tflite> --output_file=<path/to/quant_model_int_io_type.tflite> --input_type=INT8 --output_type=INT8
> ```
> 
> You can change 'INT8' to 'UINT8' as well.
> 
> If you are using `tf-nightly`, you can directly modify the input and output type of quantized models during conversion, by using this code: https://www.tensorflow.org/lite/performance/post_training_quantization#integer_only
> 
> The exception to this is quantize-aware trained models. If the model is quantize-aware trained, then in order to modify the input and output type, you need to first convert the model to a quantized tflite model (which will always have float input and ouytput) and then use the `bazel` command provided above to modify the input and output type to `tf.int8` or `tf.uint8` .

The correct command is:
```
$ bazel run -c opt //tensorflow/lite/tools/optimize/python:modify_model_interface -- --input_file=<path/to/quant_model.tflite> --output_file=<path/to/quant_model_int_io_type.tflite> --input_type=INT8 --output_type=INT8
```

**Comment by EscVM** [comment]
@MeghnaNatraj @suharshs 

Hi! I followed your guide and even after changing input output of the model I get the following error:

"
Edge TPU Compiler version 2.1.302470888
Invalid model: bin/srmodel_full_integer_qaware.tflite
Model not quantized
"

Inside the model I have an Upsamping2D layer not annotated with 'tfmot.quantization.keras.quantize_annotate_layer()'. Could be due to that?

**Comment by pl-ang** [comment]
> > @LLNLanLeN As @Namburger pointed out, you should be able to modify the model input and output by running
> > ```
> > $ bazel run -c opt //tensorflow/lite/tools/optimize/python:modify_python_inferace -- --input_file=<path/to/quant_model.tflite> --output_file=<path/to/quant_model_int_io_type.tflite> --input_type=INT8 --output_type=INT8
> > ```
> > 
> > 
> > You can change 'INT8' to 'UINT8' as well.
> > If you are using `tf-nightly`, you can directly modify the input and output type of quantized models during conversion, by using this code: https://www.tensorflow.org/lite/performance/post_training_quantization#integer_only
> > The exception to this is quantize-aware trained models. If the model is quantize-aware trained, then in order to modify the input and output type, you need to first convert the model to a quantized tflite model (which will always have float input and ouytput) and then use the `bazel` command provided above to modify the input and output type to `tf.int8` or `tf.uint8` .
> 
> The correct command is:
> 
> ```
> $ bazel run -c opt //tensorflow/lite/tools/optimize/python:modify_model_interface -- --input_file=<path/to/quant_model.tflite> --output_file=<path/to/quant_model_int_io_type.tflite> --input_type=INT8 --output_type=INT8
> ```

I got ERROR: Skipping '//tensorflow/lite/tools/optimize/python:modify_python_inferace': no such target '//tensorflow/lite/tools/optimize/python:modify_python_inferace': target 'modify_python_inferace' not declared in package 'tensorflow/lite/tools/optimize/pyth, Do you know how to solve it? Thanks a lot!

**Comment by MeghnaNatraj** [comment]
@pl-ang Could you re-run the command in comment? https://github.com/tensorflow/model-optimization/issues/431#issuecomment-647763584

(There was a typo in the command which was pointed out by @EscVM. I have updated my original comment as well.
You need to modify `modify_python_inferace ` to `modify_python_interface`  (as you can see I had misspelt `interface`))

**Comment by EscVM** [comment]
@pl-ang Moreover, remember that you have to specify the full path to indicate the input and output file!

**Comment by pl-ang** [comment]
@MeghnaNatraj and @EscVM Thank you so much for your answers! I already converted the input and output types to INT8 successfully. 

However, when I use the GPU backend on the quantized MobileNetV1 tflite model by using the TensorFlow Lite delegate APIs on Android, it shows `Internal error: Failed to apply delegate: Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors.` 

Do you know any reasons that may cause it? For the Android CPU part, the quantized model works well.

**Comment by LLNLanLeN** [comment]
@pl-ang I'm not exactly sure the issue, but it sounds like the solution is to make sure that the graph you export is static-size tensors. And you can do this by using the following line (make sure to use the correct variable name), It was recommended above:

```
q_aware_model.input.set_shape((1,) + q_aware_model.input.shape[1:])
```

**Comment by pl-ang** [comment]
@LLNLanLeN Got it!

Do you know any method to convert the tensor type from int8 to uint8? Because I found another error in Android:

```
Following operations are not supported by GPU delegate:
    CONV_2D: Max version supported: 2. Requested version 3.
    DEPTHWISE_CONV_2D: Max version supported: 2. Requested version 3.
    MEAN: OP is supported, but tensor type isn't matched!
    PAD: Max version supported: 1. Requested version 2.
    RESHAPE: OP is supported, but tensor type isn't matched!
    SOFTMAX: Max version supported: 1. Requested version 2.
```
I think it probably caused by the tensor type based on [https://github.com/tensorflow/tensorflow/issues/32468].  Do you have any suggestions? Thanks!

**Comment by MeghnaNatraj** [comment]
@pl-ang could you create a separate issue to debug this error as it's unrelated? We can further debug it in that issue.

**Comment by srjoglekar246** [comment]
@pl-ang How are you running the model? If you are using Java API and want to run a quantized model w/ GPU, look at [this](https://www.tensorflow.org/lite/performance/gpu_advanced#running_quantized_models_experimental).

**Comment by LLNLanLeN** [comment]
@pl-ang unfortunately I'm not very familiar with the error you're having. 

To convert Uint8 input, you can just use the same script and identify that input and output type is Uint8. But the problem is that, if your model is QAT, then I believe it's int8 by default. Hence if you force the input and output to be UInt8, you'll still see the `Quantize` and `Dequantize` nodes at the input and output

**Comment by pl-ang** [comment]
@srjoglekar246 I run the model by Basic usage Android (Java). I am trying to do that by the Advanced usage as below:

```
// NEW: Prepare GPU delegate with feature turned on.
GpuDelegate delegate = new GpuDelegate(new GpuDelegate.Options().setQuantizedModelsAllowed(true));

Interpreter.Options options = (new Interpreter.Options()).addDelegate(delegate);
```

But it has `cannot resolve method 'setQuantizedModelsAllowed(boolean)'` error.

**Comment by srjoglekar246** [comment]
You might need to use a newer version of the AAR/binary.

**Comment by pl-ang** [comment]
@srjoglekar246 Sorry, I am a little bit confused about it. Do you mean the dependencies? if not, could you make an example or explanation? Thanks!

The following are the dependencies I am using. 

```
dependencies {
    implementation fileTree(dir: 'libs', include: ['*.jar'])
    implementation 'com.android.support:appcompat-v7:28.0.0'
    implementation 'com.android.support.constraint:constraint-layout:1.1.3'
    testImplementation 'junit:junit:4.12'
    androidTestImplementation 'com.android.support.test:runner:1.0.2'
    androidTestImplementation 'com.android.support.test.espresso:espresso-core:3.0.2'
    implementation 'org.tensorflow:tensorflow-lite:+'
    implementation 'org.tensorflow:tensorflow-lite:0.0.0-nightly'
    implementation 'org.tensorflow:tensorflow-lite-gpu:0.0.0-nightly'
}
```

**Comment by srjoglekar246** [comment]
[This](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/gpu/java/src/main/java/org/tensorflow/lite/gpu/GpuDelegate.java#L73) is the method on `GpuDelegate`.

I am not an expert with Android packaging, but is the `lite:+` rule implementation required? Looks like the GPU delegate documentation requires only [two](https://www.tensorflow.org/lite/performance/gpu#step_2_edit_appbuildgradle_to_use_the_nightly_gpu_aar)

**Comment by pl-ang** [comment]
> [This](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/gpu/java/src/main/java/org/tensorflow/lite/gpu/GpuDelegate.java#L73) is the method on `GpuDelegate`.
> 
> I am not an expert with Android packaging, but is the `lite:+` rule implementation required? Looks like the GPU delegate documentation requires only [two](https://www.tensorflow.org/lite/performance/gpu#step_2_edit_appbuildgradle_to_use_the_nightly_gpu_aar)

Thank you so much! The quantized model works well now with GPU!

**Comment by srjoglekar246** [comment]
Awesome. Thanks for verifying! Feel free to close the issue.

**Comment by Mattrix00** [comment]
> Hi, we are actively working on integer input and output support in the conversion for QAT. (The lack of this support is why you are getting spurious dequantize and quantize operations that edgetpu doesn't like). We will update this issue when that is ready.
> 
> If this is blocking you, we have a workaround script that can change the input and output types for you to uint8, which EdgeTPU requires: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/python/modify_model_interface.py
> 
> Hope that helps!

Any news about support for QAT model ?
To run the script is necessary to build tf from source ??

**Comment by LLNLanLeN** [comment]
@Mattrix00  I believe the only way to run the script is to build from source ( I built from sources)

**Comment by corvoysier** [comment]
This issue has always been about the spurious Dequantize/Cast/Quantize in the middle of the converted model and never about the lack of support for integer inputs/outputs that I was well aware of at the time I created it.

The real fix to my issue was to statically set the input shape to (1, w, h, c), as I posted here: https://github.com/tensorflow/model-optimization/issues/431#issuecomment-646539366

I think it would be better to close it now, and maybe create a specific issue for people having issues with the script that @Namburger and @LLNLanLeN mentioned to use integer inputs and outputs (which again was NOT my issue in the first place).

**Comment by MeghnaNatraj** [comment]
The issue is resolved with TF 2.4 (the current default TF version in colab). From the [gist](https://gist.github.com/corvoysier/e4d9c6b96b103928544a57054917aace) in [comment1]( https://github.com/tensorflow/model-optimization/issues/431#issue-641064608)
![mnist_integer tflite](https://user-images.githubusercontent.com/8790823/103025491-edeba400-4577-11eb-8423-b9a81fa2b85a.png)
